File: apps/backend/src/cerebrum/api/schemas/health.py
"""Health, liveness, and readiness response shapes.

See docs/architecture/specification/38_Observability.md's Health Checks
table: three distinct check types with distinct failure semantics.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from cerebrum.utils.clock import utcnow

ComponentStatusValue = Literal["healthy", "degraded", "unavailable", "not_configured"]


class ComponentStatus(BaseModel):
    """One subsystem's contribution to the aggregate detailed-health
    status — see docs/architecture/specification/38_Observability.md
    ("Every one of the fifteen high-level components ... that has an
    external dependency ... contributes to the detailed Health check's
    aggregate status"). ``not_configured`` is this milestone's status for
    every datastore, since no client is established yet.
    """

    name: str
    status: ComponentStatusValue
    detail: str | None = None


class HealthResponse(BaseModel):
    """``GET /health`` — the detailed check, surfaced to a dashboard, not
    acted on directly by an orchestrator.
    """

    status: Literal["healthy", "degraded", "unhealthy"]
    version: str
    environment: str
    uptime_seconds: float
    timestamp: datetime = Field(default_factory=utcnow)
    components: list[ComponentStatus]


class LivenessResponse(BaseModel):
    """``GET /live`` — process is running and not deadlocked. Failure
    triggers an orchestrator restart.
    """

    status: Literal["alive"] = "alive"
    timestamp: datetime = Field(default_factory=utcnow)


class ReadinessResponse(BaseModel):
    """``GET /ready`` — process can currently serve traffic. Failure
    removes it from load-balancer rotation without a restart.
    """

    status: Literal["ready", "not_ready"]
    timestamp: datetime = Field(default_factory=utcnow)

File: apps/frontend/lib/api/health.ts
/** Health API — mirrors apps/backend/src/cerebrum/api/health.py. Root-level
 * (not under /api/v1) and unwrapped (no SuccessResponse envelope). */

import { API_BASE_URL } from "@/lib/api/client";

export type ComponentHealthStatus =
  "healthy" | "degraded" | "unavailable" | "not_configured";

export interface ComponentStatus {
  name: string;
  status: ComponentHealthStatus;
  detail: string | null;
}

export interface HealthResponse {
  status: "healthy" | "degraded" | "unhealthy";
  version: string;
  build_commit: string;
  build_time: string;
  environment: string;
  uptime_seconds: number;
  timestamp: string;
  components: ComponentStatus[];
}

/** Root URL, not `/api/v1` — health.py is mounted at the app root. */
const HEALTH_BASE_URL = API_BASE_URL.replace(/\/api\/v1\/?$/, "");

export async function getHealth(): Promise<HealthResponse> {
  const response = await fetch(`${HEALTH_BASE_URL}/health`);
  if (!response.ok) throw new Error(`Health check failed: ${response.status}`);
  return (await response.json()) as HealthResponse;
}
